File: akka-essentials/Chapter11/AkkaWithZeroMQ/src/main/java/org/akka/essentials/zeromq/java/example3/client/ClientActor.java
package org.akka.essentials.zeromq.java.example3.client;

import akka.actor.ActorRef;
import akka.actor.Cancellable;
import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.util.ByteString;
import akka.zeromq.Connect;
import akka.zeromq.Listener;
import akka.zeromq.SocketOption;
import akka.zeromq.ZMQMessage;
import akka.zeromq.ZeroMQExtension;
import scala.concurrent.duration.Duration;

public class ClientActor extends UntypedActor {
    public static final Object TICK = "TICK";
    int count = 0;
    Cancellable cancellable;
    ActorRef reqSocket = ZeroMQExtension.get(getContext().system())
            .newReqSocket(
                    new SocketOption[]{new Connect("tcp://127.0.0.1:1237"),
                            new Listener(getSelf())});
    LoggingAdapter log = Logging.getLogger(getContext().system(), this);

    @Override
    public void preStart() {
        cancellable = getContext()
                .system()
                .scheduler()
                .schedule(Duration.apply(1, "second"),
                        Duration.apply(1, "second"),
                        getSelf(),
                        TICK, getContext().dispatcher(), ActorRef.noSender());
    }

    @Override
    public void onReceive(Object message) throws Exception {
        if (message.equals(TICK)) {
            // send a message to the replier system
            reqSocket.tell(ZMQMessage.withFrames(ByteString.fromString("Hi there! ("
                    + getContext().self().hashCode() + ")->")),getSelf());
            count++;
            if (count == 10)
                cancellable.cancel();
        } else if (message instanceof ZMQMessage) {
            ZMQMessage m = (ZMQMessage) message;
            String mesg = m.frame(0).toString();

            log.info("Received msg! {}", mesg);
        }
    }
}

File: akka-essentials/Chapter11/AkkaWithZeroMQ/src/main/scala/org/akka/essentials/zeromq/example3/client/ClientActor.scala
package org.akka.essentials.zeromq.example3.client
import akka.actor.actorRef2Scala
import akka.actor.Actor
import akka.actor.ActorLogging
import akka.zeromq.Connect
import akka.zeromq.Listener
import akka.zeromq.SocketType
import akka.zeromq.ZMQMessage
import akka.zeromq.ZeroMQExtension
import akka.actor.Cancellable
import scala.concurrent.duration._

case class Tick()

class ClientActor extends Actor with ActorLogging {
  val reqSocket = ZeroMQExtension(context.system).newSocket(SocketType.Req, Connect("tcp://127.0.0.1:1234"), Listener(self))

  var count = 0
  var cancellable: Cancellable = null
  override def preStart() {
    cancellable = context.system.scheduler.schedule(1 second, 1 second, self, Tick)
  }

  def receive: Receive = {
    case Tick =>
      count += 1
      val payload = "Hi there! (" + context.self.hashCode() + ")->"
      reqSocket ! ZMQMessage(payload)
      if (count == 5) {
        cancellable.cancel()
      }
    case m: ZMQMessage =>
      val mesg = m.frame(0)

      log.info("recieved msg! {}", mesg)
  }
}